File: src/main/java/org/kobe/xbot/Utilities/ClientStatistics.java
package org.kobe.xbot.Utilities;

import com.sun.management.OperatingSystemMXBean;
import java.lang.management.ManagementFactory;
import java.lang.management.MemoryMXBean;
import java.lang.management.RuntimeMXBean;
import java.lang.management.ThreadMXBean;
import java.net.InetAddress;
import java.net.UnknownHostException;

import org.kobe.xbot.Utilities.Entities.XTableClientStatistics;

public class ClientStatistics {
    private long freeMemoryMB;
    private long usedMemoryMB;
    private long maxMemoryMB;
    private double processCpuLoadPercentage;
    private int availableProcessors;
    private long totalThreads;
    private long nanoTime;
    private String health;
    private String ip;
    private String hostname;
    private String processId;
    private String javaVersion;
    private String javaVendor;
    private String jvmName;
    private String version;
    private String type;
    private String UUID;
    private int bufferSize;
    private int maxBufferSize;

    private static final MemoryMXBean memoryMXBean = ManagementFactory.getMemoryMXBean();
    private static final OperatingSystemMXBean osMXBean = (OperatingSystemMXBean) ManagementFactory.getOperatingSystemMXBean();
    private static final ThreadMXBean threadMXBean = ManagementFactory.getThreadMXBean();
    private static final RuntimeMXBean runtimeMXBean = ManagementFactory.getRuntimeMXBean();

    public ClientStatistics() {
        this.type = "JAVA";
        this.nanoTime = System.nanoTime();
        this.maxMemoryMB = osMXBean.getTotalPhysicalMemorySize() / (1024 * 1024);
        this.freeMemoryMB = osMXBean.getFreePhysicalMemorySize() / (1024 * 1024);
        this.usedMemoryMB = maxMemoryMB - freeMemoryMB;
        this.processCpuLoadPercentage = osMXBean.getSystemCpuLoad() * 100;
        this.availableProcessors = osMXBean.getAvailableProcessors();
        this.totalThreads = threadMXBean.getThreadCount();
        this.ip = Utilities.getLocalIPAddress();

        // Retrieve hostname
        String localHostname;
        try {
            localHostname = InetAddress.getLocalHost().getHostName();
        } catch (UnknownHostException e) {
            localHostname = "Unknown Host";
        }
        this.hostname = localHostname;

        // Process and Java environment information
        this.processId = runtimeMXBean.getName().split("@")[0];
        this.javaVersion = System.getProperty("java.version");
        this.javaVendor = System.getProperty("java.vendor");
        this.jvmName = System.getProperty("java.vm.name");

        if (usedMemoryMB <= maxMemoryMB * 0.5 && processCpuLoadPercentage < 50) {
            this.health = HealthStatus.GOOD.name();
        } else if (usedMemoryMB <= maxMemoryMB * 0.6 && processCpuLoadPercentage < 70) {
            this.health = HealthStatus.OKAY.name();
        } else if (usedMemoryMB <= maxMemoryMB * 0.7 && processCpuLoadPercentage < 85) {
            this.health = HealthStatus.STRESSED.name();
        } else if (usedMemoryMB <= maxMemoryMB * 0.85 && processCpuLoadPercentage < 95) {
            this.health = HealthStatus.OVERLOAD.name();
        } else {
            this.health = HealthStatus.CRITICAL.name();
        }
    }

    public XTableClientStatistics.ClientStatistics toProtobuf() {
        XTableClientStatistics.ClientStatistics.Builder builder = XTableClientStatistics.ClientStatistics.newBuilder();
        builder.setNanoTime(this.nanoTime)
                .setMaxMemoryMb(this.maxMemoryMB)
                .setFreeMemoryMb(this.freeMemoryMB)
                .setUsedMemoryMb(this.usedMemoryMB)
                .setProcessCpuLoadPercentage(this.processCpuLoadPercentage)
                .setAvailableProcessors(this.availableProcessors)
                .setTotalThreads(this.totalThreads)
                .setIp(this.ip)
                .setUuid(this.UUID)
                .setHostname(this.hostname)
                .setProcessId(this.processId)
                .setJavaVersion(this.javaVersion)
                .setJavaVendor(this.javaVendor)
                .setJvmName(this.jvmName)
                .setHealth(XTableClientStatistics.HealthStatus.valueOf(this.health));

        // Optional fields
        if (this.version != null) {
            builder.setVersion(this.version);
        }
        builder.setType(this.type);
        if (this.UUID != null) {
            builder.setUuid(this.UUID);
        }
        builder.setBufferSize(this.bufferSize);
        builder.setMaxBufferSize(this.maxBufferSize);

        return builder.build();
    }

    public static ClientStatistics fromProtobuf(XTableClientStatistics.ClientStatistics protobuf) {
        ClientStatistics stats = new ClientStatistics();
        stats.nanoTime = protobuf.getNanoTime();
        stats.maxMemoryMB = protobuf.getMaxMemoryMb();
        stats.freeMemoryMB = protobuf.getFreeMemoryMb();
        stats.usedMemoryMB = protobuf.getUsedMemoryMb();
        stats.processCpuLoadPercentage = protobuf.getProcessCpuLoadPercentage();
        stats.availableProcessors = protobuf.getAvailableProcessors();
        stats.totalThreads = protobuf.getTotalThreads();
        stats.ip = protobuf.getIp();
        stats.hostname = protobuf.getHostname();
        stats.processId = String.valueOf(protobuf.getProcessId());
        stats.javaVersion = protobuf.getJavaVersion();
        stats.javaVendor = protobuf.getJavaVendor();
        stats.jvmName = protobuf.getJvmName();
        stats.health = protobuf.getHealth().name();

        // Optional fields
        if (protobuf.hasVersion()) {
            stats.version = protobuf.getVersion();
        }
        stats.type = protobuf.getType();
        if (protobuf.hasUuid()) {
            stats.UUID = protobuf.getUuid();
        }
        stats.bufferSize = protobuf.getBufferSize();
        stats.maxBufferSize = protobuf.getMaxBufferSize();

        return stats;
    }

    public ClientStatistics setVersion(String version) {
        this.version = version;
        return this;
    }

    public ClientStatistics setUUID(String UUID) {
        this.UUID = UUID;
        return this;
    }

    public ClientStatistics setBufferSize(int bufferSize) {
        this.bufferSize = bufferSize;
        return this;
    }

    public ClientStatistics setMaxBufferSize(int maxBufferSize) {
        this.maxBufferSize = maxBufferSize;
        return this;
    }

    public enum HealthStatus {
        GOOD, OKAY, STRESSED, OVERLOAD, CRITICAL, UNKNOWN
    }

    // Getters (same as your original code)
    public String getIp() { return ip; }
    public String getHostname() { return hostname; }
    public long getNanoTime() { return nanoTime; }
    public long getFreeMemoryMB() { return freeMemoryMB; }
    public long getMaxMemoryMB() { return maxMemoryMB; }
    public double getProcessCpuLoadPercentage() { return processCpuLoadPercentage; }
    public int getAvailableProcessors() { return availableProcessors; }
    public long getTotalThreads() { return totalThreads; }
    public String getHealth() { return health; }
    public String getProcessId() { return processId; }
    public String getJavaVersion() { return javaVersion; }
    public String getJavaVendor() { return javaVendor; }
    public String getJvmName() { return jvmName; }

    public String getUUID() {
        return UUID;
    }
}

File: src/main/python/JXTABLES/ClientStatistics.py
import os
import platform
import psutil
import socket
import time
from enum import Enum


try:
    # Package-level import
    from . import ClientStatistics_pb2 as cspf
except ImportError:
    # Standalone script import
    import ClientStatistics_pb2 as cspf


class HealthStatus(Enum):
    GOOD = "GOOD"
    OKAY = "OKAY"
    STRESSED = "STRESSED"
    OVERLOAD = "OVERLOAD"
    CRITICAL = "CRITICAL"
    UNKNOWN = "UNKNOWN"


class ClientStatistics:
    def __init__(self):
        self.nano_time = time.time_ns()
        self.max_memory_mb = psutil.virtual_memory().total // (1024 * 1024)
        self.free_memory_mb = psutil.virtual_memory().available // (1024 * 1024)
        self.used_memory_mb = self.max_memory_mb - self.free_memory_mb
        self.process_cpu_load_percentage = psutil.cpu_percent(interval=0.1)
        self.available_processors = os.cpu_count()
        self.total_threads = len(psutil.pids())
        self.ip = self.get_local_ip_address()
        self.hostname = self.get_hostname()
        self.process_id = os.getpid()
        self.python_version = platform.python_version()
        self.python_vendor = platform.python_compiler()
        self.jvm_name = platform.python_implementation()
        self.health = self.calculate_health()

        # Optional fields
        self.version = None
        self.type = "PYTHON"
        self.uuid = None
        self.buffer_size = None
        self.max_buffer_size = None

    def set_version(self, version):
        self.version = version
        return self

    def set_uuid(self, uuid):
        self.uuid = uuid
        return self

    def set_buffer_size(self, buffer_size):
        self.buffer_size = buffer_size
        return self

    def set_max_buffer_size(self, max_buffer_size):
        self.max_buffer_size = max_buffer_size
        return self

    @staticmethod
    def get_local_ip_address():
        try:
            return socket.gethostbyname(socket.gethostname())
        except socket.error:
            return "Unknown IP"

    @staticmethod
    def get_hostname():
        try:
            return socket.gethostname()
        except socket.error:
            return "Unknown Host"

    def calculate_health(self):
        if self.used_memory_mb <= self.max_memory_mb * 0.5 and self.process_cpu_load_percentage < 50:
            return HealthStatus.GOOD
        elif self.used_memory_mb <= self.max_memory_mb * 0.6 and self.process_cpu_load_percentage < 70:
            return HealthStatus.OKAY
        elif self.used_memory_mb <= self.max_memory_mb * 0.7 and self.process_cpu_load_percentage < 85:
            return HealthStatus.STRESSED
        elif self.used_memory_mb <= self.max_memory_mb * 0.85 and self.process_cpu_load_percentage < 95:
            return HealthStatus.OVERLOAD
        else:
            return HealthStatus.CRITICAL

    # Method to convert the ClientStatistics object to a Protobuf message
    def to_protobuf(self):
        stats = cspf.ClientStatistics()
        stats.nano_time = self.nano_time
        stats.max_memory_mb = self.max_memory_mb
        stats.free_memory_mb = self.free_memory_mb
        stats.used_memory_mb = self.used_memory_mb
        stats.process_cpu_load_percentage = self.process_cpu_load_percentage
        stats.available_processors = self.available_processors
        stats.total_threads = self.total_threads
        stats.ip = self.ip
        stats.hostname = self.hostname
        stats.process_id = str(self.process_id)
        stats.lang_version = self.python_version
        stats.lang_vendor = self.python_vendor
        stats.jvm_name = self.jvm_name
        stats.health = cspf.HealthStatus.Value(self.health.name)

        # Optional fields
        stats.version = self.version if self.version else ""
        stats.type = self.type
        stats.uuid = self.uuid if self.uuid else ""
        stats.buffer_size = self.buffer_size if self.buffer_size else 0
        stats.max_buffer_size = self.max_buffer_size if self.max_buffer_size else 0

        return stats

    # Method to create a ClientStatistics object from a Protobuf message
    @staticmethod
    def from_protobuf(stats_pb):
        stats = ClientStatistics()
        stats.nano_time = stats_pb.nano_time
        stats.max_memory_mb = stats_pb.max_memory_mb
        stats.free_memory_mb = stats_pb.free_memory_mb
        stats.used_memory_mb = stats_pb.used_memory_mb
        stats.process_cpu_load_percentage = stats_pb.process_cpu_load_percentage
        stats.available_processors = stats_pb.available_processors
        stats.total_threads = stats_pb.total_threads
        stats.ip = stats_pb.ip
        stats.hostname = stats_pb.hostname
        stats.process_id = stats_pb.process_id
        stats.python_version = stats_pb.lang_version
        stats.python_vendor = stats_pb.lang_vendor
        stats.jvm_name = stats_pb.jvm_name
        stats.health = HealthStatus(stats_pb.health)

        # Optional fields
        stats.version = stats_pb.version if stats_pb.version else None
        stats.type = stats_pb.type
        stats.uuid = stats_pb.uuid if stats_pb.uuid else None
        stats.buffer_size = stats_pb.buffer_size if stats_pb.buffer_size else None
        stats.max_buffer_size = stats_pb.max_buffer_size if stats_pb.max_buffer_size else None

        return stats
